Adding the Jobs API

moufmouf opened the pull request:
Hey guys,

Here is a PR to add support for the Jobs API (https://docs.gitlab.com/ce/api/jobs.html)

I wrote this PR because I needed a way to download artifacts files. This should replace the PR #164 that is based on old APIs methods that are no more documented.

One important point: since php-gitlab-api uses Buzz (until #191 is closed), I used Buzz to fetch the files. Buzz does not have the notion of streams... so the whole artifacts file is loaded into a PHP string. This is ok for relatively small files, but if you have an artifacts file that is a gigabyte large, this will certainly be a problem.

Should php-gitlab-api migrate to http-client (please do!), I highly recommend adding additional methods in the Jobs class like: `artifactsAsStream` that returns a stream instead of a string. This way, we can cope with very large artifacts files too.

Note: I've been testing this PR in [the washingmachine](https://github.com/thecodingmachine/washingmachine/pull/9) and it seems to be working.

Do not hesitate if you have any comments.

fbourigault commented:
Thanks for this PR David! As HTTPlug support has now landed in master. Could you rebase this branch and use HTTPlug instead of Buzz?

moufmouf commented:
Hey Fabien!

Woot! HTTPlug in master, this is absolutely great news!

I've rebased this PR to include HTTPlug.
Since we have support for PSR-7 responses, I also changed the artifacts download method. Now, they are returning a `StreamInterface` rather that a string, which is way more robust regarding memory usage if the artifacts files are big.

Let me know if you have any comments!

fbourigault commented on lib/Gitlab/Api/Jobs.php:
```diff
@@ -0,0 +1,133 @@
+     * @param array $scope
+     * @return mixed
+     */
+    public function jobs($project_id, array $scope = [])
```
The name `all` seems to be more used than the plural noun form. Could you change the method name to `all`?

fbourigault commented on lib/Gitlab/Model/Pipeline.php:
```diff
@@ -0,0 +1,49 @@
+    {
+        $pipeline = new static($project, $data['id'], $client);
+
+        return $job->hydrate($data);
```
Should be `$pipeline`.

fbourigault commented on lib/Gitlab/Model/Project.php:
```diff
@@ -1092,4 +1092,49 @@ public function contributors()
+     * @param array $scopes
+     * @return Job[]
+     */
+    public function jobs(array $scopes)
```
Could you add a default value for the `$scopes` argument?

moufmouf commented:
Done! Thanks for the thorough code review.

fbourigault commented:
Thank you!
